sky/jobs/scheduler.py

```diff
@@ -0,0 +1,218 @@
+    start the jobs. If there aren't enough resources, the next controller to
+    exit and free up resources should call schedule_step().)
+
+    This uses subprocess_utils.launch_new_process_tree() to start the controller
+    processes, which should be safe to call from pretty much any code running on
+    the jobs controller. New job controller processes will be detached from the
+    current process and there will not be a parent/child relationship - see
+    launch_new_process_tree for more.
+    """
+    try:
+        # We must use a global lock rather than a per-job lock to ensure correct
+        # parallelism control.
+        # The lock is not held while submitting jobs, so we use timeout=1 as a
+        # best-effort protection against the race between a previous
+        # schedule_step() releasing the lock and a job submission. Since we call
+        # schedule_step() after submitting the job this should capture
+        # essentially all cases.
+        # (In the worst case, the skylet event should schedule the job.)
+        with filelock.FileLock(os.path.expanduser(_MANAGED_JOB_SUBMISSION_LOCK),
+                               timeout=1):
+            while _can_schedule():
+                maybe_next_job = _get_next_job_to_start()
+                if maybe_next_job is None:
+                    # Nothing left to schedule, break from scheduling loop
+                    break
+
+                managed_job_id, dag_yaml_path = maybe_next_job
+                run_cmd = (
+                    f'{constants.ACTIVATE_SKY_REMOTE_PYTHON_ENV};'
+                    f'python -u -m sky.jobs.controller {dag_yaml_path} --job-id {managed_job_id}'
+                )
+
+                state.set_submitted(
+                    job_id=managed_job_id,
+                    # schedule_step() only looks at the first task of each job.
+                    task_id=0,
+                    # We must call set_submitted now so that this job is counted
+                    # as launching by future scheduler runs, but we don't have
+                    # the callback_func here. We will call set_submitted again
+                    # in the jobs controller, which will call the callback_func.
+                    callback_func=lambda _: None)
+
+                logs_dir = os.path.expanduser(
+                    managed_job_constants.JOBS_CONTROLLER_LOGS_DIR)
+                os.makedirs(logs_dir, exist_ok=True)
+                log_path = os.path.join(logs_dir, f'{managed_job_id}.log')
+
+                pid = subprocess_utils.launch_new_process_tree(
+                    run_cmd, log_output=log_path)
+                state.set_job_controller_pid(managed_job_id, pid)
+
+    except filelock.Timeout:
+        # If we can't get the lock, just exit. The process holding the lock
+        # should launch any pending jobs.
+        pass
+
+
+@contextlib.contextmanager
+def schedule_active_job_launch(is_running: bool):
```

Comment:
I found this function may overcomplicated the problem here. For launches, we do not need to preserve a FIFO order. 
Instead, we can use the `schedule_step` above to preserve the order for the order of the jobs to be scheduled, while using a semaphore to limit the total number of launches.
Reason: I think we should limit the actual launches in a finer granularity:  https://github.com/skypilot-org/skypilot/blob/83b23258f9546d64ba74dd1961282022da6e9591/sky/jobs/recovery_strategy.py#L311-L321
Otherwise, when there is bad resource capacity, and the job went into an hour-long waiting loop, it will block other jobs from being able to launch their resources.

Reply:
With this the `schedule_step` only needs to check the threshold for the total number of jobs that can run in parallel, but don't need to check the total launches.

Reply:
Updated how this works significantly. New version will transition to ALIVE (does not count towards launching limit) while in backoff.